### DISCUSSION TITLE
Feature Request - Time inside chat bubble

### DISCUSSION THREAD

**Discussion by rmControls** [flyerhq/flutter_chat_ui#96] [Ideas & Requests]
Hi,

first of all congratulations on developing this great package.... 

I have couple of questions... 
1) How mark messages deliver and read.....
2) How to customize shape and color of Chat bubble to match my theme....
3) Customize Bottom chat Send Bar and ....
4) Is there a way to group messages by date similar to WhatsApp with Date banner...?

I followed example code from here and didn't find much helpful information related to customization....
[https://pub.dev/packages/flutter_chat_ui/example](https://pub.dev/packages/flutter_chat_ui/example)

**Comment by demchenkoalex** [comment]
Hello, thank you :)

0. We have a complete documentation as well as every property in the code has documentation comments :) https://docs.flyer.chat/flutter/chat-ui/
1. Every message has status property (https://pub.dev/documentation/flutter_chat_types/latest/flutter_chat_types/Status-class.html). Statuses are controlled by the package user.
2. Theme class (https://docs.flyer.chat/flutter/chat-ui/themes and https://pub.dev/documentation/flutter_chat_ui/latest/flutter_chat_ui/ChatTheme-class.html). `primaryColor` will be the color of your messages and `secondaryColor` will be the color of received messages. Bubble shape it not customisable ATM and it is requested in the other issue already.
3. Colors/border radius/icons of the bottom bar are in the theme (https://pub.dev/documentation/flutter_chat_ui/latest/flutter_chat_ui/ChatTheme-class.html), texts are in the localization (https://docs.flyer.chat/flutter/chat-ui/localization). We don't plan any additional customizations at this point.
4. Messages are already grouped by date. The date divider will appear on every new day (midnight) and if there are more than 15 minutes difference between 2 messages. Taken from Facebook Messenger.

**Comment by rmControls** [comment]
Thank you for quick response....

is there a way to reduce a minute difference to 5 or 2 min...??

**Comment by demchenkoalex** [comment]
No, but I can add. Why you would need it every 2 minutes if I may ask? :) almost every message will have a date header which will take some usable screen space :)

**Comment by rmControls** [comment]
Yes, you are right... it doesn't make sense to add at every 2 - 5 min... 

However, It will be nice to show time at the end of chat inside/outside the chat bubble and only group the chat bubbles by date (similar to WhatsApp).... 

![image](https://user-images.githubusercontent.com/27656141/122481571-e8a66980-cf9c-11eb-8bd0-207ea715f00b.png)


Please excuse me if I am asking more...

**Comment by demchenkoalex** [comment]
Well it's just another style. We chose style similar to Facebook Messenger because of it's cleanliness (no time in each bubble) and the time is shown in the header (see screenshot in README). In the above case all 3 messages will have a time divider because of 15 minutes difference and it is small enough to not get lost in time, you see the header "10:15" and a bunch of messages, meaning this were one complete conversation and then you see "11:05" and another bunch of messages, clearly separated from the previous one. We understand that there are **a lot** of different approaches and making everything inside one library is a bit too much. For sure I don't plan to play games in calculating message size in order to decide whether I need to render time in one line or in the other, but maybe in the future we will consider optionally adding time to each message inside the bubble, but rather strictly below the content, which in our opinion will look bad, which is again the reason we went "Messenger" way.

**Comment by rmControls** [comment]
> a lot of different approaches and making everything inside one library is a bit too much

yup, I agree with that.... thanks for the explanation.... 



> but maybe in the future we will consider optionally adding time to each message inside the bubble,

thanks for considering it future development... just make sure when you display a time in bubble a group banner should be date not the time....

**Comment by alejandrocarlos** [comment]
Thank you so much for your contributions! We're making full use of all the features you've built. 
I would also like to second this request, my app needs to have visible timestamps on the messages (first responder type stuff). 

Maybe it can be a boolean or possibly allow customizing the message bubble entirely regardless of the message type?

**Comment by demchenkoalex** [comment]
Proper implementation of this feature will take a lot of time and effort so it is very low priority ATM. Just adding whatever breaks our commitment to do a good design. Bubble customisation, on the other hand, is something I would like to implement in the near future, but no ETA yet.

**Comment by OXOTHuK1976** [comment]
Will be great to have builders for different message types and have full control over bubble widget.  Thank you for your work.

**Comment by demchenkoalex** [comment]
That is exactly what I planned to do, yes.

**Comment by hibbert-de** [comment]
i used `buildCustomMessage` to implement send date/time.
in `metadata` specify if it is a _chatMessage_ or a _systemMessage_ and you can add what ever you want ;)

**Comment by demchenkoalex** [comment]
#228
